### PR TITLE
pipes - anders and angela - scrabble

### PR DESCRIPTION
**Pull request by anderschenders**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |  Because scoring is a general concept that applies to all instances versus just one instance. We're only addressing behaviour here, rather behaviour and state.  |
|   Describe an example of a test you wrote for a _nominal case_. | All tests that check for the correct return object data type (e.g., String, Integer). 	|
|   Describe an example of a test you wrote for an _edge case_. | All tests that raise argument errors for incorrect arguments are testing edge cases.  	|
|   What was your final test coverage percentage?	|  31.35%, however, we were confused by this, as after looking through the reports, we had tested for lines that were being indicated as not having been tested. Also, the coverage report is scoring 2 or our 3 spec files. 	|
|   Is there a specific piece of code you'd like feedback on?	| We would like feedback on the highest_score_from and tie class methods in the Scoring class. We worked on trying to incorporate enumerables in an attempt to make our logic more concise, but were unable to do this. 	|

### CODE REVIEW

File: lib/scoring.rb [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+      3 => ["B", "C", "M", "P"],
+      4 => ["F", "H", "V", "W", "Y"],
+      5 => ["K"],
+      8 => ["J", "X"],
+      10 => ["Q", "Z"]
+    }
+
+    def initialize
+
+    end
```
Comment (droberts-sea) [inline lib/scoring.rb:15]: Since this class is never instantiated, you don't need to define the constructor.

File: lib/scoring.rb [hunk 2]
```diff
@@ -0,0 +1,88 @@
+module Scrabble
+  class Scoring
+    SCORE_CHART = {
+      1 => ["A", "E", "I", "O", "U", "L", "N", "R", "S", "T"],
+      2 => ["D", "G"],
+      3 => ["B", "C", "M", "P"],
```
Comment (droberts-sea) [inline lib/scoring.rb:6]: This data structure certainly works, but it means you need a double loop in `score` below. It might be a little cleaner to have the map go the other way, that is, from letters to score. Then finding a letter's score is as concise as `SCORE_CHART[letter]`.

File: lib/scoring.rb [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        total_score += 50
+      end
+
+      return total_score
+    end
+
+    def self.tie(best_word, word)
+      if best_word.length == 7 #then best_word remains the same
+        return best_word
+      elsif word.length == 7
```
Comment (droberts-sea) [inline lib/scoring.rb:48]: Good job breaking this logic out into its own method!

File: lib/scoring.rb [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        best_word = word
+        return word
+      else #neither of the words are 7 characters long
+        if (best_word.length <=> word.length) == 1 #best_word is more letters
+          best_word = word
+          return word
+        elsif (best_word.length <=> word.length) == -1 #best_word is less letters
```
Comment (droberts-sea) [inline lib/scoring.rb:55]: For these, you could just say

```ruby
if best_word.length > word.length
  return word
else
  return best_word
end
```

File: specs/scoring_spec.rb [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+      test_word = "WORD"
+      test_score = 4 + 1 + 1 + 2
+      @test_class_method.score(test_word).must_equal test_score
+    end
+
+    it "Returns a 50 point bonus when input is 7 letters" do
+      test_word = "WORDERS"
+      test_score = 4 + 1 + 1 + 2 + 1 + 1 + 1 + 50
+      @test_class_method.score(test_word).must_equal test_score
+    end
```
Comment (droberts-sea) [inline specs/scoring_spec.rb:56]: What about an empty word, or an 8-letter word?

File: specs/scoring_spec.rb [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+      test_word_2 = "DO" #score = 3
+      @test_class_method.tie(test_word, test_word_2).must_equal "GO"
+    end
+
+    it "Returns the seven letter word, in the case of a tie" do
+      test_word = "AEIOULN" #score = 7
+      test_word_2 = "KD" #score = 7
+      @test_class_method.tie(test_word, test_word_2).must_equal "AEIOULN"
+    end
+
```
Comment (droberts-sea) [inline specs/scoring_spec.rb:114]: I would like to see more coverage here. For example:
- What happens if you have two 7-letter words?
- What happens if you switch the order of the words?

File: lib/player.rb [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    def draw_tiles(tile_bag)#in wave_3_game passes an instance of TileBag stored in instance variable @tilebag of Game class
+      tiles_needed = 7 - @tiles.length
+      drawn_tiles = tile_bag.draw_tiles(tiles_needed)
+      @tiles += drawn_tiles
+      return @tiles
+    end
+
+    def won?
+      if @total_score > 100
+        return true
```
Comment (droberts-sea) [inline lib/player.rb:64]: Since `<` always returns either `true` or `false` already, this could be shortened to:

```ruby
def won?
  return @total_score > 100
end
```

File: specs/scoring_spec.rb [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    it "Returns a String" do
+      test_arr = ["A", "THE", "ZZZ"]
+      @test_class_method.highest_score_from(test_arr).must_be_instance_of String
+    end
+
+    it "Returns the word in the array with the highest score" do
+      test_arr = ["A", "THE", "ZZZ"]
+      @test_class_method.highest_score_from(test_arr).must_equal "ZZZ"
+    end
```
Comment (droberts-sea) [inline specs/scoring_spec.rb:81]: What happens if you run the method with an empty array, or an array with only one word?

You should probably also check that ties are broken here - it doesn't need to be comprehensive, since you do that below, but at least one test wouldn't hurt.

File: specs/player_spec.rb [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+      test_players_tiles = @test_ob.draw_tiles(@test_tile_bag)
+      test_word = test_players_tiles[0]
+
+      @test_ob.play(test_word)
+      @test_ob.plays.must_equal [test_word]
+    end
+
+    it "Returns false if player has already won" do
+
+      2.times do
```
Comment (droberts-sea) [inline specs/player_spec.rb:59]: Good test case! This is an important behavior for the game, but it would be an easy one to miss.

File: specs/tilebag_spec.rb [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+describe "Tilebag" do
+
+  before do
+
+    @test_ob = Scrabble::Tilebag.new
+
+    default_tiles = {
+      A: 9,
+      B: 2,
+      C: 2,
```
Comment (droberts-sea) [inline specs/tilebag_spec.rb:12]: Re-implementing this hash here is pretty redundant, and is subject to the same problems that you had putting it in `TileBag` in the first place, especially since you probably just copy-pasted it. Instead, it's best to trust that you got the numbers right, and just check that, for example, the total number of tiles is correct.

File: specs/tilebag_spec.rb [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    end
+
+    it "Returns the inputed argument number of tiles " do
+      @test_ob.draw_tiles(3).length.must_equal 3
+    end
+
+    it "Removes the inputed argument number of tiles from @default_tiles" do
+      @test_ob.draw_tiles(10)
+      @test_ob.tiles_remaining.must_equal (@total_num_tiles - 10)
+    end
```
Comment (droberts-sea) [inline specs/tilebag_spec.rb:82]: You're missing a couple of interesting edge cases here:
- What happens if you attempt to draw 0 tiles?
- What happens when you attempt to draw more tiles than are in the bag?

**Comment by droberts-sea** [comment]
# Scrabble
## What We're Looking For

|  Feature  |   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions  |  yes  |
|   Both Teammates contributed to the codebase	|  yes  |
|   Regular Commits with meaningful commit messages  |  mostly - Angela, it looks like you had a few big commits, while Anders had many smaller ones. Make sure that you're remembering to stop and commit frequently. If you're having trouble breaking out of the flow, [this article might be helpful to you](https://dev.to/rpalo/plan-your-commits).  |
|  **Scoring class**	|   	|
|  `score` method	|  yes  |
|   Uses appropriate data structure to store the letter score	|  see inline  |
|   Appropriately handles edge cases	|  yes  |
|   Tests `score` edge cases (one-letter word, 7 letter word)	|  mostly - see inline  |
|  `highest_scoring_word` method	|  yes  |
|   Appropriately handles edge cases	|  yes  |
|   Tests `highest_scoring_word` edge cases (ties, empty array)	|  mostly - see inline  |
|  **Player class**	|   	|
|  Uses an array to keep track of words played	|  yes  |
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	|  yes  |
|  Returns `true` or `false` from `won?` method |  yes  |
|  Tests edge cases on the `play(word)` method for words that cannot be scored |  yes  |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	|  yes  |
|  `draw_tiles` method uses the parameter to draw a set of random tiles |  yes  |
| `tiles_remaining` method returns a count	|  yes  |
|  Tests `draw_tiles` edge cases for different parameter inputs	|  mostly - see inline  |

Good work - this code is well-organized and easy to read!

There are a few places where you've got holes in your test coverage, which I've tried to outline below. Aside from random error cases, the pattern I'm seeing is that you should make sure to explore both sides of any edge cases. For example, in `TileBag#draw_tiles`, while you consider what happens if you attempt to draw more tiles than are in the bag, you don't check that you can draw the last tile in the bag. You can look at your code and say "of course that would work", but until you test it you can't know for sure.

Becoming a master tester is not easy, and this is the first project where you've had to build tests around complex logic, so don't worry if it's not yet where we want it to be. We'll have plenty of time to practice going forward. Do make sure to spend time thinking about it for Hotel, particularly around how ranges of dates interact, and let one of us know if you want us to check your work.
